### DISCUSSION TITLE
Side-by-side form elements in otherwise vertical layout form

### DISCUSSION THREAD

**Discussion by levans** [jtdaugherty/brick#516] [Q&A]
Is it possible to combine a few FormFieldState items on the same line in an otherwise vertical arrangement?  It looks like setFieldConcat might be involve in this, but I'm not sure how. 

I currently have:
```
newForm [ label "My Callsign" @@=
                   editField mycall CallSignField (Just 1) initCallsign validateCallsign renderTextUpper id
               , label "SSID" @@=
                   editField myssid SSIDField (Just 1) initSSID validateSSID renderText id
               , label "Interface" @@= vLimit 5 @@=
                   listField (const devices) serialDevice renderListItem 1 InterfaceField
               ]
```

I'd prefer for the CallSignField and SSIDField to be placed on the same line, with a separating simple text widget " - " between them.

**Comment by jtdaugherty** [comment]
I believe you probably need to use [`setFormConcat`](https://hackage.haskell.org/package/brick-2.4/docs/Brick-Forms.html#v:setFormConcat) - `setFormConcat hBox f`.

**Reply by jtdaugherty** [reply]
Although if you want extra bits between them, then you need to add that where you want it with `@@=`, i.e. `(\w -> str " - " <+>) @@= ...`.

**Comment by levans** [comment]
Thanks.  
So, if I want the form layout to go vertically, but then shift to horizontally for a line of widgets/fields (the essential point being that this line contains multiple fields), then resume vertically again for further items; am I composing FormFieldStates, or am I composing Forms somehow?

**Reply by jtdaugherty** [reply]
I don't think that's possible in the current API. You get one way to concatenate the whole list. The only other concatenation (layout) control you have is for fields with multiple sub-fields such as radio buttons. Perhaps you could implement a custom form field to get control over that, but I don't know if it would be worth the effort. This is likely one of those scenarios where the Form API just doesn't give enough control to do something super custom. It is intended for more typical use cases where a little bit of customization is all that's needed.

**Comment by levans** [comment]
Ah, got it.  OK, at least you've confirmed that I'm probably not missing something.   Thanks again.

**Reply by jtdaugherty** [reply]
I'm happy to help, although I know in this case I wasn't able to give you a solution to this particular problem. I hope brick is working well for your project otherwise!
